src/Style.hs

```diff
@@ -300,130 +314,145 @@ initSpec = StySpec { spType = C.PointT,
-matchSel sel spec = all test (zip args patterns) &&
-                selTyp sel == spType spec &&
-                length args == length patterns
-    where
-        patterns = selPatterns sel
-        args = spArgs spec
-        -- dummies = selIds sel
-        test (a, p) = case p of
-            RawID i -> a == i
-            WildCard _ -> True
-
--- |  'procBlock' is called by 'getDictAndFns'. 'getDictAndFns' would fold this function on a list of blocks, a.k.a. a Style program, and accumulate objective/constraint functions, and a dictionary of geometries to be rendered.
--- TODO refactor procBlock and factor out helper functions
-procBlock :: (Autofloat a) =>
-    (StyDict, [(ObjFnOn a, Weight a, [Name], [a])], [(ConstrFnOn a, Weight a, [Name], [a])])
-    -> Block
-    -> (StyDict, [(ObjFnOn a, Weight a, [Name], [a])], [(ConstrFnOn a, Weight a, [Name], [a])])
-procBlock (dict, objFns, constrFns) (selectors, stmts) =
-          (newDict, objFns ++ newObjFns, constrFns ++ newConstrFns)
-    where -- Organized from used first to last
-        allVars :: M.Map Name Name
-        allVars = M.fromList $ zip k k where k = M.keys dict
 
-        -- Give specs (objects) in dict that are selected by the block's selectors
-        selectedSpecs :: [[(VarMap, StySpec)]]
-        selectedSpecs = map
-            (\sel -> let styspecs = matchingSpecs dict sel -- defined externally above procblock
-                         varmaps = map (getVarMap sel) styspecs in zip varmaps styspecs) selectors
+-- Gather a list of objective or constraint functions by mapping over mergedMaps
+-- genFns :: (VarMap -> [t] -> Stmt -> [t]) -> VarMap -> [t]
+genFns f stmts varmap = foldl (f varmap) [] stmts
 
+-- | 'procBlock' is called by 'getDictAndFns'. 'getDictAndFns' would fold this function on a list of blocks, a.k.a. a Style program, and accumulate objective/constraint functions, and a dictionary of geometries to be rendered.
+-- TODO: first split `stmts` into assigns, objs, and contrs and call procConstrFn and procObjFn with `map`?
+procBlock :: (Autofloat a) =>
+    StyContext a ->
+    Block ->
+    StyContext a
+-- Case 1: @global@ block - only objectives and constraints are allowed, and
+--- there is no need to compute any variable map
+procBlock (dict, objFns, constrFns) ((Selector C.AllT []) : [], stmts) =
+    let keys          = M.keys dict
+        identityMap   = M.fromList $ zip keys keys
+        newObjFns     = concatMap (procObjFn identityMap [])    stmts
+        newConstrFns  = concatMap (procConstrFn identityMap []) stmts
+    in (dict, objFns ++ newObjFns, constrFns ++ newConstrFns)
+
+-- Case 2: only one selector, no need to compute combinations.
+-- this function will process: (1) objectives, (2) constraints and (3)
+-- assignments
+procBlock (dict, objFns, constrFns) (selector : [], stmts) =
+    let mapsAndSpecs  = dict `matchWith` selector
+        newDict       = foldl (addShapes stmts) dict $ mapsAndSpecs
```

Comment:
looks like we’re matching the sub dict with the selector, but then adding shapes to the dict? is it possible that near the beginning of a style program, we might try to match on something that hasn’t been added to the dict yet?

Reply:
No. If you look for `initDict`, you will see we add all Substance IDs __before__ the first call of `procBlock`